### ISSUE TITLE
Not removing YouTube Music songs added to library

### ISSUE THREAD

**Issue by nckgrzn** [apastel/ytmusic-deleter#1]
First off, thanks for creating this! It works swimmingly for removing your uploaded songs, but unfortunately, it's erroring out when trying to remove YouTube Music songs. Here's the full error note:
Retrieving all library albums...
Traceback (most recent call last):
  File "/Users/nick/Library/Python/3.7/bin//ytmusic-deleter", line 10, in <module>
    sys.exit(cli())
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusic_deleter/cli.py", line 137, in remove_library
    library_albums = youtube_auth.get_library_albums(sys.maxsize)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusicapi/mixins/library.py", line 92, in get_library_albums
    albums = parse_albums(results['items'], False)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusicapi/parsers/library.py", line 55, in parse_albums
    album['trackCount'] = nav(data, SUBTITLE3).split(' ')[0]
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusicapi/parsers/utils.py", line 75, in nav
    raise err
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusicapi/parsers/utils.py", line 69, in nav
    root = root[k]
IndexError: list index out of range

**Comment by apastel** [comment]
So I looked into this, I think this is an issue with the metadata for one or more of your albums. Unfortunately since the error is happening in the `ytmusicapi` package which I didn't write, I can't quickly fix it. Basically what I believe to be happening is that one of your albums is missing a field (I think it might be "year" or "trackCount") and the `ytmusicapi` does not cleanly handle this and fails instead.

Since I can't replicate your problem unless I were to use your cookie (probably not a good idea), if I can get you to modify one of the files on your system, we can tell it to skip the broken album and print which one it is at the same time.

Open `/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusicapi/parsers/library.py`

Scroll down to line 55, which should say: `album['trackCount'] = nav(data, SUBTITLE3).split(' ')[0]`
Replace that line with these 4 lines.
```
            try:
                album['trackCount'] = nav(data, SUBTITLE3).split(' ')[0]
            except Exception:
                print(f"Failed on {album['artists']} - {album['year']}")
```
Make  sure the indentation is correct since Python relies on spacing instead of brackets for block (not sure how familiar you are with Python. So that whole `else:` block altogether will look like this:
```
        else:
            album['artists'] = nav(data, SUBTITLE)
            album['year'] = nav(data, SUBTITLE2)
            try:
                album['trackCount'] = nav(data, SUBTITLE3).split(' ')[0]
            except Exception:
                print(f"Failed on {album['artists']} - {album['year']}")

        albums.append(album)
```

Then run it again and see what happens. Thanks for your help and patience. This will help make the tool handle many different types of metadata problems.

**Comment by nckgrzn** [comment]
Soo your thoughts were correct, and I was able to clear my library. Unfortunately, I got impatient and intervened, so I can't say for sure if the fix worked completely.

I swapped out the code as directed, and ran the script again yielding results below:

[2020-09-13 20:40:46] Retrieving all library albums...
Failed on Album - 2020
[2020-09-13 20:40:52] Retrieving all leftover library songs...


















[2020-09-13 20:45:12] Processing album: Dominic Fike - What Could Possibly Go Wrong

As you can see, after some time, it began processing other albums. While the program was stalling for those five minutes though, I decided to manually remove albums with 2020 metadata. Only then did the program kick back into gear. Did the program just need some time or was it the help from a friend? Hard to say. For the sake of bug-fixing, sorry I didn't wait, but thanks again for this solution. Very much appreciated!

**Comment by nckgrzn** [comment]
Actually, the program stopped just now before completing:

Traceback (most recent call last):
  File "/Users/nick/Library/Python/3.7/bin//ytmusic-deleter", line 10, in <module>
    sys.exit(cli())
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusic_deleter/cli.py", line 146, in remove_library
    albums_removed += remove_library_albums_by_song(filtered_songs, progress_bar)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusic_deleter/cli.py", line 163, in remove_library_albums_by_song
    if remove_album(song["album"]["id"]):
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusic_deleter/cli.py", line 170, in remove_album
    album = youtube_auth.get_album(browseId)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusicapi/mixins/browsing.py", line 391, in get_album
    data = nav(response, FRAMEWORK_MUTATIONS)
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusicapi/parsers/utils.py", line 75, in nav
    raise err
  File "/Users/nick/Library/Python/3.7/lib/python/site-packages/ytmusicapi/parsers/utils.py", line 69, in nav
    root = root[k]
KeyError: 'frameworkUpdates'

Albums Processed  21%|███▉              |  565/2638 [08:57<32:50, 1.05 albums/s]

It's strange though because I have no albums left in my Albums tab, but loads of songs in my Songs tab still. I'm thinking the issue is at least partly on YouTube Music's end.

**Comment by apastel** [comment]
tl;dr: Pull the latest code -- I just pushed an update that should help it skip this "broken" song as well.

So the program works by first removing all your albums, then after that it removes all songs that don't have any album metadata. So that's why you have no albums left in your Albums tab, but a lot of songs in your Songs tab -- because it stopped working on the latter piece. But in order to save time, for each leftover song I tell it to look up the corresponding album and just remove the whole album instead of one song at a time. But maybe there are songs in your library that are somehow just not part of any album. I guess they will be revealed by seeing which ones are left over after you run the command again.

You have a lot more music in your library than I do, so you have a much bigger/varied set of data and it's not surprising that there are some tracks in your library that look a bit different in terms of back-end metadata. 

To answer your question from your previous message -- yes the program probably was busy gathering your entire library (which is large) and that's why it looked like it was stalling. I will probably update it to give some indication that it is working and not hung.

**Comment by nckgrzn** [comment]
Aside from 9 albums, which I just removed manually, it worked! YTM is still showing I have some songs from a handful of artists under the Artist tab, but when I click the Artist, nothing appears. I'm chalking that up to YTM just being a little screwy though. I can't believe they're really making us switch over to this service... Anyway, thanks again!

**Comment by apastel** [comment]
Yeah there's definitely some screwiness. I actually just tried `remove-library` again and I ended up with a leftover album by Five Finger Death Punch (I don't recall adding this lol) that couldn't be retrieved by ytmusicapi. It shows up has having exclamation marks over the song titles in the YT Music Android app as well. At least I have a piece of erroneous data to test with now.

**Comment by apastel** [comment]
So, to summarize for myself:

There are two issues discovered in this thread.

1. IndexError when calling get_library_albums() in `ytmusicapi` and there's an album that blows up at the line
`album['trackCount'] = nav(data, SUBTITLE3).split(' ')[0]` in parsers/library.py because the data being parsed doesn't have an item at the index being accessed (4) in `nav()`. This aborts the entire `get_library_albums()` operation and should instead be handled gracefully with a pull request in [ytmusicapi](https://github.com/sigma67/ytmusicapi/). The most I can do in my repo is make a change to catch the exception so the program can at least report what happened and move on to removing singles, though none of the full albums will have gotten removed. I will close this issue once I make that change here. Unfortunately I haven't replicated this yet on my own.

- [ ] Catch exceptions in `remove_library()` for each ytmusicapi call.
- [ ] Open an issue in [ytmusicapi](https://github.com/sigma67/ytmusicapi/) to handle un-parseable albums gracefully.

2. KeyError when calling get_album() in `ytmusicapi` on an album that is "broken". I can replicate this problem thanks to a Five Finger Death Punch album that somehow got transferred over from my GPM library (it's not in my GPM library and I'm not even that into them, I swear). When debugging `get_album()` on this album's ID, I end up with a 403 Forbidden error response, and that is what get_album() ends up blowing up on while trying to parse. While that should also be handled better in ytmusicapi, I can at least catch this exception (fixed in commit fc80fca82b47d49292b44008f1fb475af6c4b052) and the thread of execution can continue on to the next album so it doesn't abort the whole process. The album that couldn't be retrieved will be left over in the user's library and can be removed manually.

Both issues were discovered using ytmusicapi v0.9.2
